fix(EntityOwnerPicker): fix entity owner picker behavior (#18710)

Jona-Soerensen opened the pull request:
## Hey, I just made a Pull Request!

This PR should fix the EntityOwnerPicker bug described in the issue #18710
Based on the comments of the linked issue, I adjusted the `availableOwners` in the EntityOwnerPicker.tsx file to return only the related OwnerEntities after every EntityKind change.

- Updated EntityOwnerPicker Component to dynamically update the available Owners based on the selected Entity kind.
- Introduced `selectedKind` Prop: Added a new (optional) selectedKind prop to the useFetchEntities hook to specify the kind of entities to fetch.
- refactored useFacetEntities Hook: Refactored to return EntityOwners whenever there’s a change in the selected Entity kind.

I encountered some challenges in adapting the tests to the new logic. As a result, a few tests have been temporarily skipped. Any help is welcome here :)


https://github.com/user-attachments/assets/6b7748af-78e1-4766-9f40-e1b5148a06c2

<!-- Please describe what you added, and add a screenshot if possible.
     That makes it easier to understand the change so we can :shipit: faster. -->

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

freben commented:
Note that tests are failing.

Jona-Soerensen commented:
Updated the code in order to pass all tests.
I had to update two already existing tests to check only for `toHaveBeenCalledWith` rather than `toHaveBeenLastCalledWith` because of the implemented updateFilter reset.

Jona-Soerensen commented:
@freben @vinzscam 
Just checking to see if there’s anything else needed from my side for this PR, or if we’re just waiting for the next steps. Let me know! 😊

freben commented on plugins/catalog-react/src/components/EntityOwnerPicker/EntityOwnerPicker.tsx:
```diff
@@ -165,7 +168,15 @@ export const EntityOwnerPicker = (props?: EntityOwnerPickerProps) => {
+        owners: undefined,
+      });
+    }
+  }, [availableOwners, updateFilters]);
```
Could also have made `availableOwners.length` the dependency here, which means that whenever the LENGTH changes (and goes to zero) the thing fires. That's safe against the reference equality changing, and also perhaps even semantically more correct

Jona-Soerensen replied:
I took your suggestion and replaced the availableOwners dependency with `availableOwners.length`

Jona-Soerensen replied:
Had to revert this one because the useEffect wont update if you want to switch from e.g. groups to users because both have avaiableOwner.length of 0.

freben commented on plugins/catalog-react/src/components/EntityOwnerPicker/useFacetsEntities.ts:
```diff
@@ -41,40 +41,63 @@ type FacetsInitialRequest = {
+          facets: ['relations.ownedBy'],
+          filter: { kind: [selectedEntityKind] },
+        });
+        if (facetResponseFiltered.facets['relations.ownedBy'].length === 0) {
```
I'm trying to wrap my head around this one. So first you query by the selected kind. And if THAT has no owners you return the empty array. Otherwise, you throw away the response and query for all kinds again? Or am I just reading it wrong?

I would have expected something along the lines of the inverse:

```ts
if (facetResponseFiltered.facets['relations.ownedBy'].length) {
  return facetResponseFiltered;
}
```

Jona-Soerensen replied:
The challenge here is twofold. First, for a given selected entity kind, I need to return **all owners** and not just a subset of the available owners. In the current implementation in Backstage, the `EntityOwnerPicker` lists all owners and not just the available owners of the selected kind. 

Second, once I’ve received a `facetResponse`, I can’t perform further local filtering by kind to reduce API calls. Essentially, I need the API call to reflect the selected entity kind, so that I get a complete list of owner references for that kind.

While I understand the suggestion of simply returning the filtered response if it has owners, the issue is that we need to ensure that the filtered response includes all the owner references.

Hope I got your question right :)

freben commented on plugins/catalog-react/src/components/EntityOwnerPicker/useFacetsEntities.ts:
```diff
@@ -41,40 +41,63 @@ type FacetsInitialRequest = {
 
-  const [facetsPromise] = useState(async () => {
-    if (!enabled) {
+  const fetchFacetsEntities = useCallback(async (): Promise<Entity[]> => {
```
It feels slightly convoluted now with a `useCallback` and `useEffect`; could this be a single `useMemo` instead that depends on `[catalogApi, selectedEntityKind, enabled]`? Which evaluates to a `Promise<Entity[]>` still, of course.

freben reviewed (commented):
Pulling in @vinzscam

vinzscam reviewed (changes requested):
can I suggest to skip the filters at call time? So whenever we detect that kind group or user has been selected we make  omit the owner filter from `queryEntities()` or `getEntities()`? That way we don't have to touch react

Jona-Soerensen commented:
> Can I suggest skipping the filters at call time? Whenever we detect that a specific group or user has been selected, we ensure not to pass down the owner filter to `queryEntities()`/`getEntities()`. This way we don't have to touch React.

When you say “skipping the filters at call time” do you mean that if we detect that a `group` or `user` kind has been selected, we completely avoid passing any filters to the api? So instead of updating the filter in React, we intercept the group and user entities (as proposed in the original bug ticket https://github.com/backstage/backstage/issues/18710#issuecomment-1663824019) and reset the filter accordingly.

My understanding is that this approach would allow us to avoid extra React state updates for the owner filter and the backend would always return all owners. However, as suggested in the bug report, we intended to intercept group and user entities directly in the EntityOwnerPicker to clear the owner filter when it doesn’t apply.

Jona-Soerensen commented:
Hey, just wanted to ping you again about this PR. @vinzscam could you please elaborate your last comment?
Please let me know if there’s anything else you need from me or any remaining changes you’d like to see. Thanks!

freben commented on plugins/catalog-react/src/components/EntityOwnerPicker/useFacetsEntities.ts:
```diff
@@ -41,19 +41,37 @@ type FacetsInitialRequest = {
   const catalogApi = useApi(catalogApiRef);
-
-  const [facetsPromise] = useState(async () => {
+  const facetsPromise = useMemo<Promise<Entity[]>>(() => {
```
```suggestion
  const facetsPromise = useMemo<Promise<Entity[]>>(async () => {
```

and then you can greatly simplify everything below

freben commented on plugins/catalog-react/src/components/EntityOwnerPicker/useFacetsEntities.ts:
```diff
@@ -41,19 +41,37 @@ type FacetsInitialRequest = {
-      .getEntityFacets({ facets: [facet] })
-      .then(response =>
-        response.facets[facet]
+    return (async (): Promise<Entity[]> => {
```
you don't need this extra wrapping now

freben commented on plugins/catalog-react/src/components/EntityOwnerPicker/useFacetsEntities.ts:
```diff
@@ -41,19 +41,37 @@ type FacetsInitialRequest = {
-        response.facets[facet]
+    return (async (): Promise<Entity[]> => {
+      try {
+        if (selectedEntityKind) {
```
no need for this extra `if`. Just make a single query always, and make it have

```ts
filter: selectedEntityKind ? { kind: selectedEntityKind } : undefined,
```

freben commented on plugins/catalog-react/src/components/EntityOwnerPicker/useFacetsEntities.ts:
```diff
@@ -41,19 +41,37 @@ type FacetsInitialRequest = {
+            filter: { kind: [selectedEntityKind] },
+          });
+
+          if (facetResponseFiltered.facets['relations.ownedBy'].length === 0) {
```
same with this `if`, the `map` below will handle it fine anyway

Jona-Soerensen commented:
Refactored useFacetsEntities & EntityOwnerPicker to perform a single API call and display only the available owners.

- Consolidated the owner retrieval logic into a single API call

- Now, the picker only shows the **available owners** (based on the API response) instead of all owners

- Adjusted the reset behavior for group and user types by changing the useEffect in the EntityOwnerPicker

Open Question: Can the OwnerPicker only dispaly the available owners? 😃

vinzscam commented:
> When you say “skipping the filters at call time” do you mean that if we detect that a group or user kind has been selected, we completely avoid passing any filters to the api? So instead of updating the filter in React, we intercept the group and user entities (as proposed in the original bug ticket https://github.com/backstage/backstage/issues/18710#issuecomment-1663824019) and reset the filter accordingly.

yes exactly, basically accordingly filtering them out in all the api calls in [useEntityListProvider.tsx](https://github.com/backstage/backstage/blob/36d9578211e6634ff42ebb7f6fa818e53226beb8/plugins/catalog-react/src/hooks/useEntityListProvider.tsx#L263).

vinzscam commented:
> Refactored useFacetsEntities & EntityOwnerPicker to perform a single API call and display only the available owners.
> 
> * Consolidated the owner retrieval logic into a single API call
> * Now, the picker only shows the **available owners** (based on the API response) instead of all owners
> * Adjusted the reset behavior for group and user types by changing the useEffect in the EntityOwnerPicker
> 
> Open Question: Can the OwnerPicker only dispaly the available owners? 😃

the `EntityOwnerPicker` displays the available owners only if the `mode` prop is set to `owners-only`.

vinzscam commented on .changeset/real-news-marry.md:
```diff
@@ -0,0 +1,5 @@
+'@backstage/plugin-catalog-react': patch
+---
+
+Fixed the `EntityOwnerPicker`component to dynamically update the `availableOwners` based on the selected Entity kind.
```
mmm I'm not sure this is the fix for #18710.
It seems the issue we want to address is that if I have already selected an owner and then switch the kind to users or groups, the owner filter shouldn't prevent me from seeing the available groups or users. However, the "available owners" feature should remain as it is.

This means that if an adopter sets `<CatalogIndexPage ownerPickerMode='all' />` in their `App.tsx`, we should display all users/groups present in the catalog. On the other hand, if they set `<CatalogIndexPage ownerPickerMode='owners-only' />`, we should only display users/groups that are assigned as the owner of at least one entity.

Jona-Soerensen replied:
Thank you. I just created a new [PR](https://github.com/backstage/backstage/pull/29678) wich just adds a filter for the "user" & "groups" entity kind inside the useEntityListProvider. This fixes the Bug without touching the state management.
https://github.com/backstage/backstage/pull/29678

Jona-Soerensen commented:
Created a new PR for the requested changes. 
https://github.com/backstage/backstage/pull/29678

camilaibs commented:
Hi @Jona-Soerensen 👋🏻 , should this pull request be closed in favor of [this](https://github.com/backstage/backstage/pull/29678) other one?

Jona-Soerensen commented:
Yes! Let’s go ahead and close this PR in favor of the other one 🚀

benjdlambert commented:
Thanks @Jona-Soerensen :tada:
